### DISCUSSION TITLE
Replace m.route by a Router component

### DISCUSSION THREAD

**Discussion by andershol** [MithrilJS/mithril.js#2925] [Ideas]
**Is this something you're interested in implementing yourself?** No

### Description
This is just a rough idea. Maybe it would be useful to instead of writing
`m.route(document.body, "/page1", { "/page1": Page1 });`
to be able to write something like:
`m.mount(document.body, { view: () => m(Router, { default: '/page1', routes: { "/page1": Page1 } }) });`
such that routing could be done inside layouts and not only at the top level. It might also allow routing to be split into components and nested.

A larger sample (the implementation is a quick hack), tested in Firefox and chrome:
* Currently the m.route function is used something like [sample 1](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4AbhgBOAAgAyGAJ60ArsWkBeacAA6abcTEQYAd0TSwytNWIR6ACjFpaAExgBKLbv37JMYssk0aSw7bRAsDAg9Si9ib1JQkEIARjCKaTCAWQVpDAAHPLC3Cl1pMvKQsPxjSXy00vLGyoIcNGV6oMau5rTguyx8SRVGfFkogGt04GlCXzBEMIB6PIwAcxhUkGl2DJAABTWYaU23YoauisTekIGh1RhRiamZuYWQZcOAJjDt9LCD9bSb4gU4lToXMo9Sh9W7DB5jNCTLQvGDzJYrdYAZh+O3+h2k2JBoNi8UhiXwQkYDF6jhcD2ohGgzl8aFObgA3LFWLpuXo0BIZACNhpPHoDEZTOZLNZbGg7B5pr5-IE+vIlKp0lC-vt8ScOdseboBdIhZ8RToxYYTGYLFYbPYFdIlQEgiE1cNNVdoXjAcDTuyDWhecahZjzbErZLbTKHcjnSq3YoPX0wr0fUdCf7A7zdLD7nZnLRqMpWsR8AAjFwKbUfdabKYk94YjZhMxC5Jg7xNr6tk1fTtxbtY3uhsGsDmUEBwGCwGNoBA8T4AdkQmIAtJ9MYgAAxsDggTA4PAUuACGj0KmiHgkMjpSvOHLTCKSVZRMzbgO8lLI5+vtDvgMVmcZwolWMxkm3PIAA8A3LDBqHGVY7jQZwzAAYmA5xP10apajyZEQLgPIoEUG1YBgwN8FaZRkSAkC0DA45IIouCEKQlQUPQmBuOwtAKQvERaIwYDQPA5jeMnadZ3tec8GSRBkgAVjYABdKgoAmBdUAPLg8CwCBiFmaB8H4ScAnIa9iGIPI4EQRZFksPJEP4rBFn0wzJGgAABT58G3fAABY3IMoyoBMs9iAUPJuCnahPLyURWBU1ggA). Here an application have three pages that share a common layout.
* If m.route was replaced by a route component, that sample might be directly translated onto [sample 2](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10RACdaAro174AVgip0GTYngBuGXgAIASoOFKAvEuAAdNLuJyIMAO6IlYAWmrEI9ABRy0tACYwAlDv1LfShcoAHlpKWPgADopwMAAKGMSEmDgOrrTUAjgM+FBp8fZoRBhwhPhwAgBGcMS8EGgA5g4ATB4eANw+fmC0yk6KSgDWSrX+Lu748dVw+PxC8F56aH5+AUpJ8CHIALrti0v+fZHEIbogAHonSgDUA9Mw4VAY1DAOAPTI+ABUlwD8pwAkwFYDg8AB9kLpdJsIRDNi86hQlAByaG6P4AMkRHg6eyUt3uj2eL0QDmQGAAtAAvAAMZIAnAB9TaXDwOFEQ-Bs3Qctn4EEeF4QCjYnFKBz0hFoBFyLyaAB8orWU3CAmKDjQHgRiIciMuTm+iPwiMQiOQpzJL02mMuiI+HkxbWFvhWvBgITQZlUMDqAFFAuEHIcNUplNoXfgYIEYNQHIEIvFCF4QSDgzBw5Ho7HDoQrkoTi8Tg7dksIGBRbx5o6-Ir8F1eN7HoQHGqERAZfLM4psHBkGhNiFeMgIJcAIybQsilPEAS8RZYJyjVMTXhTGaMbv9TYIju8LvjvasYUHouhabqVPRYjztyL4iTfDuMAYARQI5J3MgfOUVbPqAI4Au-EngsaoBBgVg918F0pxnUU50RMAwHtHZfCPI9TFqVJTHGVxXG9OQZAAGQgKomBgXhWRAcJaHCKp4hgE4ESsGw7EcGBpW8XYwhdVwd1MYEdnYSwMCgaI2koEBolgWx8gQHhGipRBhypNgOBANY8DECQaHoRhmB4FYCIwABPdQQgWQxjDMCwmOkxxnGvCsDGIQwoOnWcKKwDBagY-RDD80gKMIYcGPfABZIylAwcJwgLIVj18OcTnwUwdxiyhKwSij8EyAQfPivZEpAEK5y4s98CItB+j-JRCBdMBEDzSI6hgYKQCUQSTjiZqlFalo4onTKTmKhxStmcraiqnQarqhqPyamBGguDqQC611FpAPqMr8QrhtGxhxsq6rapgerGowZqAGYloRTrztdK6Ns2pz-MGggpF05yv3ssZqEIaAeKYFpC0MVDfLQFZVuHMzfKMExzEsaxbLQYEptcmC50MkyhARHav1u7reradr9CPCG7saaGnMs+GbJY5H5knNzQgcTH1BxiiQvxtaCyJ0Hwb6VaLspiy4esxG6ZR4BGfRlnjLZ5mhrxla7qUB6WlaYm0CPfQwiwQQGBSNIMhkfByjcIzqupiwUblZm1FmXhqofJ8XwsREXnm4dEQRVd4AsczPo9u7WosSG4r8ubyZOUPyfDwP5oemPLri1h2q8cDWnEySozp2SQEaAB2RAAA4VM4HANOoOAtPemQ8BIMgETN1wIqlzzeDqWoLCpDWjyCqb287tBu41yIcNqOoLCU8JAg18pHn6OoZjQVwLAAYhw1xe-0ZLUqm1xiPxIzrNgWfNeypgBCmseD-qKeqRnueF6X-XV6UNeYE-7eClrhhr4wced8eoPzPkeLOMApK5zwMORSABWNgm4QBQAmrJVAakuB4CwBABINQoCaXEtOcgPASDEBoogF4LxrDhEXvgOgWAXhYJwdAAAAo0fAVJ8AABYGHYNqtAfBVBiBGXCNwCS1AajhFkKwTYrAgA)
* This means that the router could be moved inside the Layout component, so the individual pages could be animated (this is my current primary use-case) as in [sample 3](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10RACdaAro174AVgip0GTYngBuGXgAIASoOFKAvEuAAdNLuJyIMAO6IlYAWmrEI9ABRy0tACYwAlDv1LfShcoAHlpKWPgADopwMAAKGMSEmDgOrrTUAjgM+FBp8fZoRBhwhPhwAgBGcMS8EGgA5g4ATB4eANw+fmC0yk6KSgDWSrX+Lu748dVw+PxC8F56aH5+AUpJ8CHIALrti0v+fZHEIbogAHonSgDUA9Mw4VAY1DAOAPTI+ABUlwD8pwAkwFYDg8AB9kLpdJsIRDNi86hQlAByaG6P4AMkRHg6eyUt3uj2eL0QDmQGAAtAAvAAMZIAnAB9TaXDwOFEQ-Bs3Qctn4EEeF4QCjYnFKBz0hFoBFyLyaAB8orWU3CAmKDjQHgRiIciMuTm+iPwiMQiOQpzJL02mMuiI+HkxbWFvhWvBgITQZlUMDqAFFAuEHIcNUplNoXfgYIEYNQHIEIvFCF4QSDgzBw5Ho7HDoQrkoTi8Tg7dksIGBRbx5o6-Ir8F1eN7HoQHGqERAZfLM4psHBkGhNiFeMgIJcAIybQsilPEAS8RZYJyjVMTXhTGaMbv9TYIju8LvjvasYUHouhabqVPRYjztyL4iTfDuMAYARQI5J3MgfOUVbPqAI4Au-EngsaoBBgVg918F0pxnUU50RMAwHtHZfCPI9TFqVJTHGVxXG9OQZAAGQgKomBgXhWRAcJaHCKp4hgE4ESsGw7EcGBpW8XYwhdVwd1MYEdnYSwMCgaJC30F4XksCBgiwQQGAsABBKAclMJQlwgcpZjgJRIh3HBhH0MJaCgVwAFk5KObQwlk6xiB2ayLJCJjbHyBxeARagEQwCtdhLUUMC0TRtGsB9ahgVwvCg6dZ3wYyzIstyPL3KKYKMkzzNsxKdH8ExzFFNtQgcTy1K8cCBNzNBKBAaJYBc+gEB4RpGkQYcADY2A4EA1jwMQJBoehGGYHgVgIjAAE91BCBZDGMMwLGcli0CvdwfMMQwUtnCisAwWoGP0NbiEMOcTkIYcGPfUyxrU8JwgLIVj18Y6CFMHdbsoStHoo-BMgEPaHr2J7zrnLiz3wIi0H6P8lEIF0wEQPNIjqGAzpAJRBJOOIkaUFGWnuidPpOIGHBB2YwdqSHsphmA4YRjAkcaC50ZATHXQZkBcY+vxAa-YHT1J8GKeAaHYfhj9EZgABmRmEQxunXSl9mOYMQ6VYJggpEGw6v2ca98GoQhoB4pgWjE4hUP2tAFvyIT3AAeSEZbPA4p0F3vWgwmoB44DgIiqmw1wKMfdx1ALZDJ2i1YPRifgsGI54rccF04GM-CfKWHWxlSMIMBwvDCOIwayIojA0AgbbFqYVxzqTlPPDDsr9CPfQVhZ4cpo6ehymp7oyJgWT8PmnOYHt4h7pm3L5usOqlvmcPUq+oOYF2r9ZaxnHWjRxvm76FnGnb3ZO+7l0XX7mBB7toQx6MCfLCnxbgWyjbCpOGsh+XmXmblpQ2baTe0CPFucsJb718IfWsvdT7n2Hpffa185q32Yq5WeT8nqv3cO-d8LMlAK1-ubAwaAVgKRuvvce8CE4z2dn4FBDhRoTRgf9OcahZjuWymsCwiIzy8ERAiB8T4XzsJeOLYc3DgxnjgBYBYE5aZIxRhYVueMRTSJgGzORctGgKJxEohWqikYSw0X4VgaMWj1y3mgBymVUjpEyMQfA5Q3BjQREQ8IbQqo1SjItBqIAAAsEtEBUg6pwHAPVqDeyqhrGQeASBkARHY1wV0hbbV4HUWoFgqQbyPKdbKiTkloFSRvSIOFah1AsMOKk4RAgb3KI8fodQZhoFcBYAAxDhVw6T9D4BehgcI2VXDEXxGNeasAKl-2+kwAQ2UCm9PqCUspwyqnUBqXUhpShGkwDWRvMAQzUlKCpGpIQtA2kFHCQwCZtA4AQEWhYF0Dw7D4Q2Vs7G2MN60HwrwTZtA8oGxwkwQ5esBoyCUPKXpcgJk5ymcU7Gsz8lnIufkCwGBKjGVmBvdCrgEgzKpAAUg3oQJedQSAYuxUoWAYBiB+I3jUfFZK0lKHmYsuSyzGkITAIc9pi9hhCxLmXPI9BEDssWPgAArHAQ5AABfoMAxpgD0usfl2VpXu2ytRR4FyBk0uqCXOAtYsCIA1WgOANznilKxRQKkuClDEFoEqyI1BVUtQ3nqrV3QdWOsNQ4M16SRmL0mpy0u5c4XeqELiYVYqJVSpldpQNRwhYKqwNalVxABnDk9Za+NtrE3krRn-VxMBaoeLwIKxAEtGhsE3CAKA5MGqoC6lwPAscEg1CgL1Kq05yA8BIMQGiiAJLWHCDUv5WAXj1phtAUVjR8BUnwF4odFyR1NvEFVRN4RuDVWoDUcIshWCbFYEAA)
* And it might even be possible to nest the Routers, so the routes could be put inside the components to which they belong as in [sample 4](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10RACdaAro174AVgip0GTYngBuGXgAJGWAA4AFXjEgAPJQF4lWfPyEx8a7XooqY6jRmKFDSgORuA3AB00C5QBKgsKuwL7exHIQMADuiEpgAmjUxBD0ABRyaLQAJjAAlEphaEqlSv5K+kYmaopwMI7OmDjpObTUAjgM+FDtTmloRBhwhPhwAgBGcMS8EGgA5ukmZoyW1hC6PUzzzvn5PiVlYLTKmYpKANZKc+XZefhOM3CmwfCFxWVlFc3wrsgAugdPqUKrViK5vCAAHqQpQAakuphgaigGGoMHSAHpkPgAFRwgD8UIAJMBWOl8gAfZDebz-Wm0-6Y+a2NwM7zEgBkbnyvmBZSRKLRGMxiHSyAwAFoAF4ABklAE4APr-OH5dLs2n4TXebWa-CU-KYiAUPn8srpJW2NC2OSFAwAPiU6R+zzUAhGLvyrPSbjhmQJbnwbkQbmQUMlmP+PLhblx+R5+zNnwq2lcaFiSgCMHmAFFdGp0mDvUplEZtPgYLoYNR0pswYRCpTKaWLFWa3XLE4XAjIZjIUnDp8IGBnbx3snga78MdeLm0YR0i7bBB7U764psHBkGh-q5eMgIHCAIz-QfmuwObvw8vIWX-ZAH+9bBbOSWnwxGNyYtxKAmlu9-jGSZplmBZ0llWwnyA2BXxcSUlGPQp4mgoFzWWV41h0DZXFUTR1n0BE8MaQg0P5bRiAEXgSiwdJgHKaI4mdNdjEyO4LEeXhnhWeBkAuf5bA3Xgt0KVhz2BVhJ0kodjBecwsL0XD7Hw7DdDI2SeLGGBiHSYjrwRLJcg44gnnwPIwAwAQoHBZslD7SFrSsqBbGAbQhXReIZgEGAxPUiiqJKJZfUTNDpOkmI5jaGIHhyHJczkGQABkIGmJgYF4DUQDUWg1GmJwYAchIkhSAZ0hgO0ijNPTnFcDw0Iw+SrFUpSHAI+qkRyYSYgpA52ASDAoHqQdKBAepYBK+gEB4Y8ACZEAADnfObjwANjYDgQB+PAxAkGh6EYZgeAqRKMAAT2CUJwkiRj4kSZJUgyQy8gnNAIgifzqNYyEsAwOYHKut7SEywhj0KyEAFlTqUDA1DUAdTRk0paMhfAYmEuHKEnT5kYILoBH+xHgRxwraIa1ZkrQC4XKUQhrEQPtanmGBQcoFRFCZ4hQyVABlGAoDAX8+shRwmcQgdvSxonMpJpY5PJuYqaKGm6YZjAmZmwriHZ7Sud5-nBdsYW1ZgJQNZAPYEYvJHpdZ0m5YsCnFfo2mdHpkBMUZmAAGZNe1zm3B5vmBaUIWQBFk2ffNi2AeICJrZRqQDtj1mnosahCGgTqmD2QcIjC8I-HOcPj0u17rtiW7ioetAKUqmSPpom3LfQ2XNMd6mXbAN3MS1iYWdsLXeA53Wg4NuyQAAFQwCYxfN1klBbNxm-5O224VjuVfd3uzYHv2R-10pQ6nmezYlwnraCcxeGpn54jcOYM14JelHMyzrLvnvp+PZ+eLgeIPitn2XuLN-4MQrsxQwTpaJuEIDNZ+bhj6IR5CHZe5ogHTzNqAqI4Da6OlYjAuBrJEFwNEqgz4rAQ57E8CHXw0kKjhxmqXCI2CmJ3QmjXd4rZKKfWJqzI2otT7UPzoXZQ4cvZMPLqwqupVOENy+iAMGYdjZKEjvsGhaBpK+AqAAQVhhIlhld7oyLrp8ORtETrnSEGQ1il9hA3y4HfV4T9bCvyspzdwHtjbfygq8P+Ji0Hu09iApQxdrGlFVurSE8QGFhPHp4pmkdonGy9tYihYlxLCJMFgQQDBWjtE6DIfAExcinVsLotQ+wRpjRrNXKaIBjwKkQMeL261OA4G2tQOAu1E4yDwCQMgthik5ChvRH6Q85jxFlEI3wIMlZjPmBMpQUylC1FinMeY8RjyyjUGpJQEw0QXHmGYNAOR4gAGJYo5GmYMNGMMlY5BSkKU6t1YC7OkvgPGStVkPIWJs7Zuz9nUEOcc05SgzkwAhdc-APSGBfIwGs35iF-nXKqXzGpAw6nHiaQAVjYAJEAUAFZTVQJtLgeAsAQGcLMKAO0RpUXIDwEgxBcqIExJiJIahDnQtoFgTEFKqXQAAAIzXwLKfAAAWPllLabQFpVQYgp01DcFGtQWYahZCsH+KwIAA). But this means that a route like "/page1" also match the route "/page1/tab1" which might not be desirable.

### Why
Sample 3 is my motivation. It seems to me that I in this case want to not have the routing on the top level, but instead inside the layout.
Some of the motivation is that it seems that the diff-algorithm will not (refer to sample 1) match up the Layout component used in Page1 to the Layout component used in Page2, so the Layout is not updated, but removed and replaced, making animation of the content imposible.

### Open Questions
It seems the implementation could be made less hackish if there was a onafterview method, that was called after all child view methods had been called, or if the view method for the child components was called by the m-method (i.e. so the view methods for the child components was called and completed inside the parent view method).
The use of target:'_Self' is a hack (that at least work in Firefox and Chrome) to have m.route.Link use the href and not its own onclick as the onclick use a new value for m.route.prefix, not the one used to set href.

**Comment by kczx3** [comment]
I believe rewriting the router to fit your idea has been discussed and hopefully is an aim for v3. Mithril desperately needs nestable routing imho

EDIT: Fixed a horrendous typo

**Comment by cavemansspa** [comment]
@andershol -- not sure if it addresses your specific issue, but you can manage page transitions at the layout level. see this [gitter message](https://gitter.im/mithriljs/mithril.js?at=5dacda84ea0d6159fd9ddd5a) for link to flems example.

re:
> Some of the motivation is that it seems that the diff-algorithm will not (refer to sample 1) match up the Layout component used in Page1 to the Layout component used in Page2, so the Layout is not updated, but removed and replaced, making animation of the content imposible.

here's your [flems augmented](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4AOmgBuGAE4ACADIYAnrQCuxOQF45wCWOKSIMAO6I5YNWmrEI9ABSS0tACYwAlLv3EDMmMTUZNDksezEQLAwINHCKbwMDUPDCAEZYuXCAWWU5DAAHPPD3OOC5MrKkghMZfNiJcoaQsIIcNDU60sbyyvTQrHwZdUZ8BWiAawpdOUI-MERwgHo8jABzGDSQOXYMkAAFVZg5Dfdi+q7u5t77fsGNGBHxyeBp2fmQJYOAJnCtyfD9tZyb4gE4lc4XcJXG5De6jNATKYzGBzRbLNYAZh+23+BzkmJBoPiPk0TXC+CEjAY6Scrnu1EI0BcfjQJ3cAG5vKwJFyYmhLNZbPQLBg3AB5DSOZxuTx6Uo0tz4Fy0frUKAYOBwUZwYj4EUuZpgEUwIZFDmlPwBIJyNCmOS7QZYCBwGD2fk2Oxoex+OC0KCSDxeTpyeX3JX9PUAUX9DC1lJgMmaGDQEEigptaBc6W9vv97LOrDzaB5Emk8gB620gbK9AARsjaH4-FhaP7zIaxRoSgYjKY21Z3Q4ZXILYFgpV8O2YNF0jjAcc2VtuSXZHavpXZdW0HWwA34zBm63hR3iF3DMYzBZ+2n7EOR1bx5Pp5QduWgabF0Xl2WDuj12da-Wjb7i2MBtka4ont4PYXm6163v4o6kgQj4xM+s6HPi7IfsWaALAswYrgAggUf64fh3bnn2AoejegZ4Y0d5jvYSiqJ2Ej0V0oQAEowjITzWlw5gAOS8UJkxuIaahQMQwkfGsKRicOMJwOYsocecqIHBs5jlikJTqV0mlrMCOlfPp+HgkZMD4qZGLmecrBbCcZrqTy9FuQs-TNlYxD2Eq1BqK0Oo1q4yiTMReSFtCdx+bQAVBfgIUuGFOxyes6Qbhp7xoulSC6MybgyOYN5aAAfE0LFDJMoS6Sc7BnIZ2VfOEqkFfGxXuGVFUqFVTTlp8dVgllaU2flTCFR1XWhJVnZ9T+dXcuylA0MqeTQPGeA1hgdbkFQzqwAOaAIDwAAsAAciAnQAnGwHAgJgOB4OSGrLRSIh4CQZCTElOTPJEMgrNE5gAAwLjyqRTP9gNoCDC7LC4LjRCs5gpMDeQAB4Llt1BjCstwZuYADECMuGDEj4NU+RTIjcB5GqyhtrAmMfvgrRqFM8OI2gyNHGjzPY7j+MuETMCiwuYBMyDcjA7kGi0GTaDkvQlKaM8eS0HAEBpuYfhqrY-ri5LRxHAuIEyBLtAXgyCNMArSvCAwcjlYjkgc3qSMo3zcMa1rHrmNtPpQHcC4mBALjEIQnvAwApAuhBTisJBR7HciwGAMmg8OECJxnWMYDjePqATciE2AZcK+TKFTEmKYYNrVf4AArHACsAAJjDAyhgDUOBwEeU7BM83fKlMtDLNQWsM5nxA1EdO4yFgiAz0mcB6y6qMxxQwNYY5xC0KP4+T4gKQLsvc8NovZ+r3XLrb2DLOTkM1fJqmfuPxochNy3H7t533dcH3d+qsLAOgPvnI+J8thyD3mAiexAp73x5MtfaMBDrHRAI3RA6JPhsAALpUCgOMY6qB7pcDwI6COMhoD4H4MtQI5AeAkGIHkFSeErB5FxkrLAnktYzGgK3T4+Bgb4BOjwyh1DaFUHgXkbgIA4DUCoXkUQrBcGsCAA)

**Comment by andershol** [comment]
> not sure if it addresses your specific issue, but you can manage page transitions at the layout level. see this [gitter message](https://gitter.im/mithriljs/mithril.js?at=5dacda84ea0d6159fd9ddd5a) for link to flems example.

Thanks for the pointer. I tried to [simplify it](https://flems.io/#0=N4Igxg9gdgzhA2BTEAucD4EMAONEBMQAaEAMwEskZUBtUKTAW2TQDoALAF0fmPSk6IBqECAC+RekxYgAKgCdMscp3LQAMpgCeEAK6dWAK2olIAoZxFJOAAk6Llq6AGVOmQTYC8N4AB0oNjbyeoKumGAA1ig2NAC6RP6BIQBGelD40XEJATbQ9kowKmpQmbH+YgDc-ogAHtgQ8rZmMLYKBUUa2iFeNgAUAJReAHw+iTYAbpjydjA9+Y7FYYJVOfKInLryAX45SVBga+6I0b3ANmDslPhrAWKDniM7gc8TU0Gz3pwwrMH6iGGRIjnTY3WzeRg-EKIVgAc3WAxWL0Ck2mUEQAHcAEpQnryGA0PGsJBQGGcdgAWgATLEbABCbxgEEWRFI8ikPporFQwaE7C6GDsXqM+Sg-oVGyIeB4d6serYBH+MYvGDwcj4RC9TnYv5Ar6sFJpfBAi5XG66755ByFJxQMVK556g26dI9OIsl6O-YQRjYazFV2xd02CT28bkDEnM4m+DXITB+4jaOxqDZF7QZKIUgNRBrRgQcbHPpRy4xm7x4ajXbPACCIu0suCnAgnC02GhfIFrBwvq0vUd+lSzqN5xLybtVcCaw2WxsnJsAAVgoxyHhems4PACxXPfNrcVZfzBeuEAX+uOXiGcpVyhU+HgkGAbdQ0ABOFAABnEkhADGYInnmBwm0CzQN8xh8GYgjCGgpDOo+-oqmqiDVvA8C9JKjAwECv6IIMTwSvAmGsFm8gAKLhIKiAVvhgTQvg3qsGAWAwDA6grgYub5hqADkiHquS5BQOS8jkDCXDcUCvGqvxITCaJ4mSXxiACUJSCkJwEk2FJSHkrJakaeezx0QxTGYCxbEtF2+D4Jq0iGXc5T+LB+w2jYSm9DCEDqJmnBAk66RAoJg4BbkXo+n60B4WMSkoWh-nDp53nqTYAD8WlKbp+jkvp3E2NE2kyVlIliQZLIxahvRBYaQKJT5qXpdJymCXJJW5flGXNTlhl4JwsjkMwIS9AMFbQJA4XrPuJHkRcQoVmAAz9EClLvu+Yo2OUd6Sog8GgSIAAcKCUgAbF+Uh-mw4GmHkFgiP19SND4NjAXunQ6PowY2KQS7rSArAAPTPR0UCaG9BjGL4IArP4EK-IIvT0YyzACKwqT4FoQIQ39ENAvhmMQ9EZw3Oq8gnAmMT2oEjC9IDNogyEOM2BEiBaNEACMwapkilO9BDHCIJgxPY1pwCw4WePiDY7D88T3GLRTzxU7zjCYIJQtnC0WhINEENS-JnAoKzK0AKQVMk4QRDCvzpOSkDwA0KAZqJiAQxz8svBDyuq8QbsK70MNQqwbFQBEDPsGspDayAf1ZhALsSD90cQLHIBnv09plFenOYzH+OVi8RM5qTFY0PLVM08UdP6AzTMszYlKuxOvu81LAs5tjPuBNxItQpHifJ2IkvSzmEkdzYisgMk+hNlAauhUx5CRNEVEPI9lwtA0Wgo+bCIffHENm5EENnpzXPj6wnsz8Qj0a1rP26yVBvG6b5uW4aNsIPb2CYEgluIEIccny5j9C+7dG4vCpv7P4gdBIh1XuHXuZt5B-VZgAhOiDkFHzPOnMYl5AiY3QSgNUucaJBCEMTSMNh3D2FmHcYupdqZWiBpXXyj0a7RAAMw2AANQ-XJC7HhVDCRqgbkAsePNfot0Ft7MBzwu6iwQVMP6hD8AQwHpI4egDT7iMnpwaes9RqqkXhKaikt2Iby3pEHeA894T3NpgtOVYM6BDEOUccEFvTYEoDmEQZsMy8BIPebaT4RCsw-KdH80gRCMRYu48w0EJ4QDRnnMeUwYSCWiO+FkX9rKCRhBklkXF5CkDtuiaIlxrJCBZPgFcvptDREtmqFk98uBsxWuMdgLJ0RqjJK0984x0QFILEUkpZS1TqigCyGEOAOHYBqDYdhszJkiXwOSQQ4Ujjv3gLoRgsA2ZfRsKzfZhz5BLLVKsxA6zBDCQgOiGA0RlxCUghYSh+gIBj2ak8gQKwXEzygKjLQNg+at2mPhBpKypj8zZjYP6BzoWwphQAFm+f4f5gKL7JLBZs7ZJR4U2CRWMTFwRSl1zhZSKpNSsC12KbUFk1KajkmqWsHaOLbbYsGTmYpNzyS13KeMppiA9bREwK85FvzWAdSEsVLgj0lD9XcMUFAEqWrStYJSGA4ofkAAEa5fWkLMJVUrbBnC+t6R6u4SKMBQLuLAcNDbviNkQVaGq7BvLOOaholrrVHF6E63e-hxWNUypwZVRrKFQDlTaRVgbZKGsBWqjV-htXM11cwfV0aip60eiaxgZqrQWqtVaG1GpfUDybLmgo+avW2uNo6ta1jFRQADTpTqdUziyuVpGpV+k43quDImnVihU1uUDS25KxrvpurzR6gtBQi29HJHah1JaXXltgJWwt3rl0-KbYVYN3a23ho7QqjKek6qqt7Vqgderh06VPWOz6E67BTvkJ6jdcNl1lsnRW6dVaNQLprVu-w4h4ggFVMHZ8dAInnRAMuMkIl4BGBMCATYvA0BcE4LgFAf0-rOmwBbRi3o-qwbDpQTVlJWDvlYAiojKgSMIcuiAFsbYRAwAOOQbAlgxCxDEEAA) a bit (to use css-transitions, and when it is assumed that there will be exactly one TransitionLayout on each page by only handling a single outbound/inbound and by starting slide in oncreate, you don't need to hook into the mithril resolveRoute, but can just update stack and check for direction in oncreate).
This handles a bit of a different part of the issue, than what I was talking about. But of course it is related and something I also need to implement. I think that approach is to simple however, as it does not seem to handle deep links and I think I want a more "app like" behavior where the back button takes you up the the navigation hierarchy, but not down or sideways (while supporting "sub tasks", I guess), so some knowledge of the navigation hierarchy is required.
So I think adding a "direction" to mithril wouldn't really make sense, but a before/after route event on the m.route would be nice. But that is another issue I guess.

> > Some of the motivation is that it seems that the diff-algorithm will not (refer to sample 1) match up the Layout component used in Page1 to the Layout component used in Page2, so the Layout is not updated, but removed and replaced, making animation of the content imposible.
>
> here's your flems augmented

So what you are saying is that the diff-algorithm handles the individual elements inside a route-resolver, but not inside a component. Thanks for pointing that out. I had not payed attention to that distinction, although in hindsight I guess I used that in another project.

**Comment by dead-claudia** [comment]
We are currently looking into this. #2281 is closed, but that's the general direction we've been looking. (Currently, our route resolvers haven't been paying off super well.)

Full component-driven routing is unlikely to occur until we get React-like context (#2148), but this is a space we've been exploring.

**Comment by kczx3** [comment]
Would you happen to have a status on where you’re at with “looking into this”?

**Comment by dead-claudia** [comment]
There's #2281 (closed for other reasons), but the current state of my design work is [here](https://github.com/isiahmeadows/mithril.js/blob/redesign-redux/design/router.md). There's no official timeline, but there's two main blockers:

1. React-like context is a necessity if you want these routes to be testable.
2. I want it to have both React Router's flexibility and Mithril's simplicity, and that's not easy at all.

**Comment by kczx3** [comment]
@isiahmeadows Thanks for pointing to this. It was an interesting read. Although, for someone being less theoretical in their thinking, it would be beneficial to have some sudo code along with the API to see how code might look using that API.

**Comment by dead-claudia** [comment]
@kczx3 Here's a couple small concrete examples: [1](https://github.com/isiahmeadows/mithril.js/blob/redesign-redux/examples/threaditjs/mithril-redesign-vanilla/app.mjs), [2](https://github.com/isiahmeadows/mithril.js/blob/redesign-redux/examples/todomvc/todomvc.mjs#L16-L21)

Neither are actually runnable - they're just examples.

**Comment by kczx3** [comment]
Thank you!
